### PR TITLE
Add GPTQ support for Gemma

### PR DESCRIPTION
**Pull request by TechxGenus**
The latest version of AutoGPTQ already supports Gemma. Add some minor modifications can provide support for vllm.

My test code is as follow:
```python
from vllm import LLM, SamplingParams

sampling_params = SamplingParams(temperature=0, max_tokens=256)
llm = LLM(model="TechxGenus/gemma-2b-GPTQ")

outputs = llm.generate(["def min(arr):\n"], sampling_params)
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(prompt + generated_text)
```

The output result is:
```python
def min(arr):
    min = arr[0]
    for i in range(1, len(arr)):
        if arr[i] < min:
            min = arr[i]
    return min


def max(arr):
    max = arr[0]
    for i in range(1, len(arr)):
        if arr[i] > max:
            max = arr[i]
    return max


def main():
    arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    print(min(arr))
    print(max(arr))


if __name__ == "__main__":
    main()
```

### CODE REVIEW

**Comment by simon-mo** [comment]
thank you for the PR. can you confirm the original model still work?

**Comment by TechxGenus** [comment]
Sure! The test code is as follows:
```python
from vllm import LLM, SamplingParams

sampling_params = SamplingParams(temperature=0, max_tokens=256)
llm = LLM(model="google/gemma-2b")

outputs = llm.generate(["def min(arr):\n"], sampling_params)
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(prompt + generated_text)
```
The output result is:
```python
def min(arr):
    min = arr[0]
    for i in range(1, len(arr)):
        if arr[i] < min:
            min = arr[i]
    return min

def max(arr):
    max = arr[0]
    for i in range(1, len(arr)):
        if arr[i] > max:
            max = arr[i]
    return max

def main():
    arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    print(min(arr))
    print(max(arr))

if __name__ == '__main__':
    main()
```

File: vllm/model_executor/models/gemma.py [hunk 1]
```diff
@@ -325,15 +325,23 @@ def load_weights(self,
...
                 break
             else:
+                # Skip loading extra bias for GPTQ models.
+                if name.endswith(".bias") and name not in params_dict:
+                    continue
                 # GemmaRMSNorm is different from Llama's in that it multiplies
                 # (1 + weight) to the output, instead of just weight.
                 if "norm.weight" in name:
                     loaded_weight += 1.0
+                if name == "lm_head.weight" and name not in params_dict:
```
Comment (esmeetu) [inline vllm/model_executor/models/gemma.py:343]: Quick question: Why this check is needed?

File: vllm/model_executor/models/gemma.py [hunk 2]
```diff
@@ -325,15 +325,23 @@ def load_weights(self,
...
+                # Skip loading extra bias for GPTQ models.
+                if name.endswith(".bias") and name not in params_dict:
+                    continue
                 param = params_dict[name]
                 weight_loader = param.weight_loader
                 weight_loader(param, loaded_weight, shard_id)
                 break
             else:
+                # Skip loading extra bias for GPTQ models.
+                if name.endswith(".bias") and name not in params_dict:
```
Comment (esmeetu) [inline vllm/model_executor/models/gemma.py:337]: Could you merge this check with the one above and place it outside the for loop?

File: vllm/model_executor/models/gemma.py [hunk 3]
```diff
@@ -325,15 +325,23 @@ def load_weights(self,
...
+                    continue
                 # GemmaRMSNorm is different from Llama's in that it multiplies
                 # (1 + weight) to the output, instead of just weight.
                 if "norm.weight" in name:
                     loaded_weight += 1.0
+                if name == "lm_head.weight" and name not in params_dict:
+                    continue
```
Comment (WoosukKwon) [inline vllm/model_executor/models/gemma.py:344]: QQ: Why is this needed?

Comment (TechxGenus) [reply vllm/model_executor/models/gemma.py:344]: Gemma's input and output weights are shared and there is no separate `lm_head` parameter.
There will be no problem when loading the original model, but when loading the quantized model, `lm_head` is not included in `params_dict`. I skip it to load the model normally.

File: vllm/model_executor/models/gemma.py (hunk 1 above)
Comment (TechxGenus) [reply vllm/model_executor/models/gemma.py:343]: Gemma's input and output weights are shared and there is no separate `lm_head` parameter.
There will be no problem when loading the original model, but when loading the quantized model, `lm_head` is not included in `params_dict`. I skip it to load the model normally.

File: vllm/model_executor/models/gemma.py (hunk 2 above)
Comment (TechxGenus) [reply vllm/model_executor/models/gemma.py:337]: I copy them from https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/llama.py#L377-L379 and https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/llama.py#L385-L387.
Modifying the `for-else` structure may cause excessive modifications. I recommend reusing existing code.

File: vllm/model_executor/models/gemma.py (hunk 1 above)
Comment (TechxGenus) [reply vllm/model_executor/models/gemma.py]: After reinstalling from the latest main branch of transformers and AutoGPTQ, I found that removing them works fine. Thanks for testing!

File: vllm/model_executor/models/gemma.py (hunk 3 above)
Comment (TechxGenus) [reply vllm/model_executor/models/gemma.py]: After reinstalling from the latest main branch of transformers and AutoGPTQ, I found that removing them works fine. Thanks for testing!

File: vllm/model_executor/models/gemma.py (hunk 2 above)
Comment (esmeetu) [reply vllm/model_executor/models/gemma.py:337]: @TechxGenus I think we can do this optimization on `Gemma` model file. Just put this check onto `for (param_name, shard_name, shard_id) in stacked_params_mapping:` without changing `for-else` code part.

**Review by esmeetu — APPROVED** [review]
Thanks for your contribution!
